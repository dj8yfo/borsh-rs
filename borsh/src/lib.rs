#![cfg_attr(not(feature = "std"), no_std)]

/*!

# Crate features

### Ecosystem features

* **std** -
  When enabled, this will cause `borsh` to use the standard library. Currently,
  disabling this feature will result in building the crate in `no_std` environment.

### Default features

* **std** - enabled by default.

### Other features

* **rc** -
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [`Rc<T>`](std::rc::Rc)/[`Arc<T>`](std::sync::Arc) respectively.
  In `no_std` setting `Rc`/`Arc` are pulled from `alloc` crate.
* **hashbrown** -
  Pulls in [HashMap](std::collections::HashMap)/[HashSet](std::collections::HashSet) when no `std` is available.
  This feature is set to be mutually exclusive with **std** feature.
* **bytes** -
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [Bytes](bytes::Bytes) and [BytesMut](bytes::BytesMut).
* **bson** -
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [ObjectId](bson::oid::ObjectId).


### Config aliases

* **hash_collections** -
  This is a feature alias, set up in `build.rs` to be equivalent to (**std** OR **hashbrown**).
  This alias gates [schema](crate::schema) and [schema_helpers](crate::schema_helpers) modules, as
  [BorshSchema](crate::schema::BorshSchema) relies on [HashMap](std::collections::HashMap) existing.
  Gates implementation of [BorshSerialize](crate::ser::BorshSerialize) and [BorshDeserialize](crate::de::BorshDeserialize)
  for [HashMap](std::collections::HashMap)/[HashSet](std::collections::HashSet).


*/

#[cfg(not(feature = "std"))]
extern crate alloc;

#[cfg(feature = "schema")]
pub use borsh_derive::BorshSchema;
pub use borsh_derive::{BorshDeserialize, BorshSerialize};

pub mod de;

// See `hash_collections` alias definition in build.rs
#[cfg(feature = "schema")]
pub mod schema;
#[cfg(feature = "schema")]
pub mod schema_helpers;
pub mod ser;

pub use de::BorshDeserialize;
pub use de::{from_reader, from_slice};
#[cfg(feature = "schema")]
pub use schema::BorshSchema;
#[cfg(feature = "schema")]
pub use schema_helpers::{try_from_slice_with_schema, try_to_vec_with_schema};
pub use ser::helpers::{to_vec, to_writer};
pub use ser::BorshSerialize;

#[cfg(all(feature = "std", feature = "hashbrown"))]
compile_error!("feature \"std\" and feature \"hashbrown\" don't make sense at the same time");

/// A facade around all the types we need from the `std`, `core`, and `alloc`
/// crates. This avoids elaborate import wrangling having to happen in every
/// module.
#[doc(hidden)]
#[cfg(feature = "std")]
pub mod __maybestd {
    pub use std::{borrow, boxed, collections, format, io, string, vec};

    #[cfg(feature = "rc")]
    pub use std::{rc, sync};
}

#[cfg(not(feature = "std"))]
pub mod nostd_io;

#[doc(hidden)]
#[cfg(not(feature = "std"))]
pub mod __maybestd {
    pub use alloc::{borrow, boxed, format, string, vec};

    #[cfg(feature = "rc")]
    pub use alloc::{rc, sync};

    pub mod collections {
        pub use alloc::collections::{btree_map, BTreeMap, BTreeSet, LinkedList, VecDeque};
        #[cfg(feature = "hashbrown")]
        pub use hashbrown::*;
    }

    pub mod io {
        pub use super::super::nostd_io::*;
    }
}
