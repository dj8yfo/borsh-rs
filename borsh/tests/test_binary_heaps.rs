#![cfg_attr(not(feature = "std"), no_std)]

#[cfg(feature = "std")]
use std::collections::BinaryHeap;

use borsh::{from_slice, BorshSerialize};

#[cfg(not(feature = "std"))]
extern crate alloc;
#[cfg(not(feature = "std"))]
use alloc::{
    collections::BinaryHeap,
    string::{String, ToString},
    vec,
};

macro_rules! test_binary_heap {
    ($v: expr, $t: ty, $snap: expr) => {
        let buf = $v.try_to_vec().unwrap();
        #[cfg(feature = "std")]
        if $snap {
            insta::assert_debug_snapshot!(buf);
        }
        let actual_v: BinaryHeap<$t> = from_slice(&buf).expect("failed to deserialize");
        assert_eq!(actual_v.into_vec(), $v.into_vec());
    };
}

macro_rules! test_binary_heaps {
    ($test_name: ident, $el: expr, $t: ty) => {
        #[test]
        fn $test_name() {
            test_binary_heap!(BinaryHeap::<$t>::new(), $t, true);
            test_binary_heap!(vec![$el].into_iter().collect::<BinaryHeap<_>>(), $t, true);
            test_binary_heap!(
                vec![$el; 10].into_iter().collect::<BinaryHeap<_>>(),
                $t,
                true
            );
            test_binary_heap!(
                vec![$el; 100].into_iter().collect::<BinaryHeap<_>>(),
                $t,
                true
            );
            test_binary_heap!(
                vec![$el; 1000].into_iter().collect::<BinaryHeap<_>>(),
                $t,
                false
            );
            test_binary_heap!(
                vec![$el; 10000].into_iter().collect::<BinaryHeap<_>>(),
                $t,
                false
            );
        }
    };
}

test_binary_heaps!(test_binary_heap_u8, 100u8, u8);
test_binary_heaps!(test_binary_heap_i8, 100i8, i8);
test_binary_heaps!(test_binary_heap_u32, 1000000000u32, u32);
test_binary_heaps!(test_binary_heap_string, "a".to_string(), String);
