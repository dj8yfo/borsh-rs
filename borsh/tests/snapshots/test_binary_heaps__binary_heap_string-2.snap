---
source: borsh/tests/test_binary_heaps.rs
expression: buf
---
[
    1,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    97,
]
